crates/kornia-pnp/src/refine.rs

```diff
@@ -0,0 +1,240 @@
+
+    while iters < params.max_iters {
+        iters += 1;
+        // Numerical Jacobian J (2N x 6)
+        let mut j = vec![0.0f32; 2 * n * 6];
+        let h_rot = 1e-4f32; // radians
+        let t_scale = x[3].abs().max(x[4].abs()).max(x[5].abs()).max(1.0);
+        let h_trans = 1e-4f32 * t_scale; // world units
+
+        for k_idx in 0..6 {
+            // Central differences
+            let h = if k_idx < 3 { h_rot } else { h_trans };
+            let mut x_plus = x;
+            let mut x_minus = x;
+            x_plus[k_idx] += h;
+            x_minus[k_idx] -= h;
+            let _ = project_all_in_place(&x_plus, &mut residuals_p);
+            let _ = project_all_in_place(&x_minus, &mut residuals_m);
+            for i in 0..(2 * n) {
+                j[i * 6 + k_idx] = (residuals_p[i] - residuals_m[i]) / (2.0 * h);
+            }
+        }
+
+        // Build normal equations: (J^T J + lambda I) delta = -J^T r
+        let mut a = [0.0f32; 36];
+        let mut b = [0.0f32; 6];
+        for r_i in 0..(2 * n) {
+            let r_val = residuals[r_i];
+            for c in 0..6 {
+                let j_ic = j[r_i * 6 + c];
+                b[c] += j_ic * r_val;
+                for d in 0..6 {
+                    a[c * 6 + d] += j_ic * j[r_i * 6 + d];
+                }
+            }
+        }
+        // Damping
+        for d in 0..6 {
+            a[d * 6 + d] += lambda;
+        }
+
+        // Solve A delta = -b
+        let mut rhs = [-b[0], -b[1], -b[2], -b[3], -b[4], -b[5]];
+        let mut a_mat = a;
+        if let Some(delta) = solve_6x6(&mut a_mat, &mut rhs) {
+            // Tentative update
+            let mut x_new = x;
+            for i in 0..6 {
+                x_new[i] += delta[i];
+            }
+            let err_sq_new = project_all_in_place(&x_new, &mut residuals_p);
+            if err_sq_new < err_sq_base {
+                // Accept step
+                x = x_new;
+                residuals.copy_from_slice(&residuals_p);
+                if (err_sq_base - err_sq_new) < params.eps {
+                    converged = true;
+                    err_sq_base = err_sq_new;
+                    break;
+                }
+                err_sq_base = err_sq_new;
+                lambda = (lambda / params.lambda_mul).max(1e-12);
+            } else {
+                // Reject step, increase damping
+                lambda *= params.lambda_mul;
+            }
+        } else {
+            // Singular system, increase damping
+            lambda *= params.lambda_mul;
+        }
+    }
+
+    // Write back results
+    rvec.copy_from_slice(&[x[0], x[1], x[2]]);
+    t.copy_from_slice(&[x[3], x[4], x[5]]);
+
+    let rmse = (err_sq_base / (2.0 * n as f32)).sqrt();
+    Ok((rmse, iters, converged))
+}
+
+// Rodrigues helpers removed; use SO3::exp for rotations.
+
+// Dense 6x6 solver using Gaussian elimination with partial pivoting.
+fn solve_6x6(a: &mut [f32; 36], b: &mut [f32; 6]) -> Option<[f32; 6]> {
```

Comment:
should we put this later in linalg? @cjpurackal

Reply:
Later, yes.